### PR TITLE
[DeepSeek-V4] Add experimental SM89/L40S FP8 support slice

### PR DESCRIPTION
**Pull request by JJ-Lin**
# DeepSeek-V4 Flash SM89/L40S support slice

## Summary

- Make the DeepSeek-V4 torch top-k fallback CUDA-graph safe by replacing Python `.any()` control flow with tensor-only masking.
- Add L40S FP8 W8A8 and Triton MoE tuned config assets for the DeepSeek-V4-Flash FP8 path.
- Document the L40S/SM89 recipe as an experimental FP8 + Triton fallback path.

Related: #23602

## Motivation

DeepSeek-V4 Flash can run on 8x L40S with the FP8 checkpoint, but the current SM89 path needs two upstreamable fixes before it can be maintained cleanly:

- the torch top-k fallback must not branch on CUDA tensors during graph capture;
- L40S tuned kernel configs should live in the normal SGLang config asset directories instead of being injected out of tree.

This PR intentionally keeps the scope small. It does not attempt to upstream the whole external sparse decode fallback or the local deployment wrapper.

## Modifications

1. `python/sglang/srt/layers/attention/dsv4/indexer.py`
   - Rewrites `topk_transform_512_pytorch_vectorized` so short-sequence handling is expressed with tensor operations.
   - Keeps the short-row padding semantics while avoiding `.any()` on CUDA tensors.

2. `test/registered/unit/layers/test_dsv4_indexer.py`
   - Covers all-short rows.
   - Covers mixed short/top-k rows.

3. L40S config assets
   - Adds observed L40S W8A8 FP8 configs for DeepSeek-V4-Flash FP8 shapes.
   - Adds Triton 3.5.1 MoE configs for `E=256,N=256`, including the `_down` variant.

4. Docs
   - Adds an experimental L40S note to the DeepSeek-V4 cookbook.

## Accuracy Tests

Local checks:

```text
git diff --check origin/main..HEAD
python3 -m py_compile \
  python/sglang/srt/layers/attention/dsv4/indexer.py \
  test/registered/unit/layers/test_dsv4_indexer.py
```

`pytest` collection on my local macOS Python is blocked by a missing optional runtime dependency (`orjson`), not by the test itself.

Previous remote/container checks from this experiment:

```text
PYTHONPATH=python pytest -q test/registered/unit/layers/test_dsv4_indexer.py
2 passed, 6 warnings in 7.80s
```

```text
loader smoke ok
```

## Speed Tests and Profiling

8x L40S deployment evidence using this config slice plus local follow-up runtime work:

```text
512-word prompt / 512 output x3:
decode_tps.mean=140.132

4K input / 1K output:
decode_tps.mean=131.299
```

## Non-goals

- No claim of complete DeepSeek official API parity.
- No DeepGEMM/DeepEP enablement on SM89.
- No upstreaming of the local SGLang Docker wrapper, auth proxy, or web test page.
- No upstreaming of the external sparse decode fallback in this PR.
- No model-alias compatibility layer for `deepseek-chat` / `deepseek-reasoner` in this PR.

## Follow-ups

- Package and review the SM89 sparse decode fallback as a separate PR.
- Decide whether the local `hash_topk` PDL guard should be upstreamed; the WIP patch is saved outside the PR-ready branch.
- Build a compatibility shim if this deployment must behave exactly like DeepSeek's public API, including official model aliases, `thinking` object normalization, Beta prefix/FIM routes, and context-cache observability.

## Checklist

- [x] The change is scoped to an SM89/L40S support slice.
- [x] Unit coverage added for the torch top-k fallback.
- [x] Config assets live in the standard SGLang config directories.
- [x] Cookbook note added for the experimental L40S path.
- [x] Registered the new unit test with the CI registry.
- [x] `git diff --check origin/main..HEAD` passes.
- [x] Python syntax check passes for the changed Python files.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** — add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** — `run-ci` is required first.

### CODE REVIEW

**Comment by JJ-Lin** [comment]
Lint is green after registering the new unit test with the CI registry. The remaining PR Test failures are from the gate requiring the `run-ci` label, and I do not have permission to add labels on this repository. Could a maintainer please add `run-ci` when ready?

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 1]
```diff
@@ -108,59 +108,46 @@ def topk_transform_512_pytorch_vectorized(
     page_bits = (page_size - 1).bit_length() if page_size > 1 else 0
     page_mask = page_size - 1
 
-    positions = (
-        torch.arange(max_seq_len, device=device).unsqueeze(0).expand(batch_size, -1)
+    sequential_indices = (
+        torch.arange(TOPK, device=device, dtype=torch.int32)
+        .unsqueeze(0)
+        .expand(batch_size, -1)
     )
```
[the remark this answers is not in the document]
Comment (JJ-Lin) [reply python/sglang/srt/layers/attention/dsv4/indexer.py]: Addressed in the latest push: `sequential_indices` now keeps shape `(1, TOPK)` and relies on broadcasting.

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 2]
```diff
@@ -108,59 +108,46 @@ def topk_transform_512_pytorch_vectorized(
...
-            torch.arange(TOPK, device=device, dtype=torch.int32)
-            .unsqueeze(0)
-            .expand(batch_size, -1)
+        batch_indices = (
+            torch.arange(batch_size, device=device).unsqueeze(1).expand(-1, TOPK)
         )
-        sequential_valid = sequential_indices < seq_lens.unsqueeze(1)
+        gathered_scores = scores[
+            batch_indices.flatten(), raw_indices.clamp(min=0).flatten()
+        ].view(batch_size, TOPK)
```
[the remark this answers is not in the document]
Comment (JJ-Lin) [reply python/sglang/srt/layers/attention/dsv4/indexer.py]: Addressed in the latest push: `gathered_scores` now uses broadcasted advanced indexing with `(batch_size, 1)` batch indices and `(batch_size, TOPK)` raw indices.

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 3]
```diff
@@ -108,59 +108,46 @@ def topk_transform_512_pytorch_vectorized(
...
 
+        valid_topk = gathered_scores != float("-inf")
+        needs_sequential = (seq_lens <= TOPK).unsqueeze(1).expand(-1, TOPK)
```
[the remark this answers is not in the document]
Comment (JJ-Lin) [reply python/sglang/srt/layers/attention/dsv4/indexer.py]: Addressed in the latest push: `needs_sequential` is now kept as `(batch_size, 1)` and broadcasted by `torch.where`.

**Comment by pakjoeng** [comment]
Any plan for support MXFP4 on SM89?
Works on SM90 #24816

**Comment by JJ-Lin** [comment]
Thanks for asking. I do not plan to claim SM89 MXFP4 support in this PR.

#24816 is an SM90/Hopper path: it wires FlashInfer's SM90 mixed-input CUTLASS MoE backend, and that route depends on Hopper-specific pieces such as the SM90 mixed-input layout/kernel path. L40S is SM89/Ada, so I do not think it can reuse that backend by just relaxing an arch guard.

For this PR I am intentionally keeping the SM89 target to the FP8 checkpoint path (`sgl-project/DeepSeek-V4-Flash-FP8`) with Triton FP8 W8A8/MoE configs plus the DSV4 top-k fallback. That is the path I have validated on L40S. MXFP4 on SM89 would be a separate follow-up that needs its own kernel/backend work and accuracy validation, not just a config change.

**Comment by Zephyr69** [comment]
Any news on merging this pull?

**Comment by YangZeyu95** [comment]
Support sm89 will be very helpful as we have 8xL40 servers. Any plan to merge this pr?

**Comment by y1ny** [comment]
Any news? Support sm89 will be very helpful.

**Comment by mengdeer589** [comment]
Support sm89 will be very helpful as we have 4x4090 48GB servers. Any plan to merge this pr?
